scipy/signal/_signaltools.py

```diff
@@ -843,6 +843,56 @@ def _calc_oa_lens(s1, s2):
     return block_size, overlap, in1_step, in2_step
 
 
+def _swapaxes(x, ax1, ax2, xp):
+    """np.swapaxes"""
+    shp = list(range(x.ndim))
+    shp[ax1], shp[ax2] = shp[ax2], shp[ax1]
+    return xp.permute_dims(x, shp)
+
+
+def _split(x, indices_or_sections, axis, xp):
+    """A simplified version of np.split, with `indices` being an list.
+    """
+    # https://github.com/numpy/numpy/blob/v2.2.0/numpy/lib/_shape_base_impl.py#L743
+    Ntotal = x.shape[axis]
+
+    # handle array case.
+    Nsections = len(indices_or_sections) + 1
+    div_points = [0] + list(indices_or_sections) + [Ntotal]    
+
+    sub_arys = []
+    sary = _swapaxes(x, axis, 0, xp=xp)
+    for i in range(Nsections):
+        st = div_points[i]
+        end = div_points[i + 1]
+        sub_arys.append(_swapaxes(sary[st:end, ...], axis, 0, xp=xp))
+
+    return sub_arys
+
+
+def xp_pad(x, pad_width, mode='constant', *, xp, **kwargs):
+    # xp.pad is available on numpy, cupy and jax.numpy; on torch, reuse
+    # http://github.com/pytorch/pytorch/blob/main/torch/_numpy/_funcs_impl.py#L2045
+    # for mode = 'constant'
+    if mode != 'constant':
+        raise NotImplementedError()
+
+    value = kwargs.get("constant_values", 0)
+
+    if is_array_api_strict(xp):
+        np_x = np.asarray(x)
+        padded = np.pad(np_x, pad_width, mode=mode, **kwargs)
+        return xp.asarray(padded)
```

Comment:
Have you considered skipping tests for array API strict instead? I don't think dispatching to NumPy for Array API strict is a useful pattern

Reply:
Uhm, a precedent of skipping  array_api_strict is going to open quite a Pandora's box. Is  *very* tempting though, NGL.

Reply:
Also the whole of array_api_strict is dispatching to numpy internally. This should not happen in user code though, which is why I'm suggesting to move this function to array-api-strict.

Reply:
> Uhm, a precedent of skipping array_api_strict is going to open quite a Pandora's box. Is _very_ tempting though, NGL.

It might have already been opened in `scipy.differentiate`...

Reply:
Yeah, since it is already done elsewhere I'm happy for you to do it here too, then we can resolve it by adding `pad` to `xpx`.

Reply:
Well, as long as it's going to end up in `array-api-extra`, as a user of `xpx` I would very much prefer that it works across all backends including `array-api-strict`.

Reply:
Ah, does this use modes of pad which look difficult to implement in array-api-extra?

Reply:
> Here though the pragmatic thing seems to be to make it just work for all xpx users. If all it takes is a single if is_array_api_strict(xp) branch, there's no reason to not do it IMO.

I definitely disagree - the whole point of array-api-strict is to mimic a dummy array library which we don't know anything about. If we are relying on features which are unimplementable without relying on array-api-strict's caveat that we can dispatch to NumPy, then we should just skip the tests instead and mark it as TODO.

Reply:
I don't know why we are even discussing this. The only implementable mode is `constant`. Whether to call `np.pad` or pad manually is frankly a minor implementation detail.

```
In [23]: import array_api_strict as xp

In [24]: a = xp.reshape(xp.arange(3*4*5), (3, 4, 5))

In [25]: padded = xp.zeros(tuple(x + 2*pad_width for x in a.shape), dtype=a.dtype)

In [26]: padded[(slice(pad_width, -pad_width, None),)*a.ndim] = a

In [27]: xp.all(padded == xp.asarray(np.pad(a, pad_width)))
Out[27]: Array(True, dtype=array_api_strict.bool)
```

Reply:
Ah cool, so this is implementable just fine in xpx for xp-strict.

I just mean that there is no need to add special-casing for array-api-strict - we can either add the full implementation for now or punt on it until it is in xpx.

Reply:
All "real" backends implement `pad` natively. So scipy should dispatch to their implementations (torch's implentation is spelled a little differently but we still should dispatch to that one). It is only array-api-strict which does not have it, so one way or the other there's some branching. 

And scipy itself should not carry backend-specific code unless _really_ necessary, so all of it should go to array-api-extra. Whether the implementation for array-api-strict uses numpy or not is up to array-api-extra devs IMO.

Reply:
Agreed. How would you like to proceed? I would suggest:
- skipping for array-api-strict for now
- making a PR to array-api-extra
- unskipping the tests once the array-api-extra PR is in and we can bump the submodule

Reply:
My ideal plan would be close, only I'd avoid the churn of skipping for now, unskipping later:

- merge this PR as is
- make a PR to array-api-extra
- remove `pad` when available from array-api-extra

An alternative is to hold off merging this one until the `array_api_extra.pad` is available,  and update it once.

This PR is not critical and does not block further adoption work.

Reply:
> An alternative is to hold off merging this one until the array_api_extra.pad is available, and update scipy once.
> 
> This PR is not critical and does not block further adoption work.

Want to go for that, then?